Kernel/Process.cpp

```diff
@@ -1551,6 +1556,127 @@ void Process::crash(int signal, u32 eip)
+    size_t offset = 0;
+
+    sti(); // Re-enable interrupts so we don't shit the bed (we need I/O)!!!
+    auto pname = String(m_name);
+    auto fd_or_error = VFS::the().open(String(String::format("%s_%u.core", pname.characters(), RTC::now())), O_CREAT, 0400, current_directory(), UidAndGid { m_uid, m_gid });
+    auto& fd = fd_or_error.value();
+
+    elf_file_header.e_ident[EI_MAG0] = 0x7f;
+    elf_file_header.e_ident[EI_MAG1] = 'E';
+    elf_file_header.e_ident[EI_MAG2] = 'L';
+    elf_file_header.e_ident[EI_MAG3] = 'F';
+    elf_file_header.e_ident[EI_CLASS] = ELFCLASS32;
+    elf_file_header.e_ident[EI_DATA] = ELFDATA2LSB;
+    elf_file_header.e_ident[EI_VERSION] = EV_CURRENT;
+    elf_file_header.e_ident[EI_OSABI] = 0; // ELFOSABI_NONE
+    elf_file_header.e_ident[EI_ABIVERSION] = 0;
+    elf_file_header.e_ident[EI_PAD + 1] = 0;
+    elf_file_header.e_ident[EI_PAD + 2] = 0;
+    elf_file_header.e_ident[EI_PAD + 3] = 0;
+    elf_file_header.e_ident[EI_PAD + 4] = 0;
+    elf_file_header.e_ident[EI_PAD + 5] = 0;
+    elf_file_header.e_ident[EI_PAD + 6] = 0;
+    elf_file_header.e_ident[EI_NIDENT] = 16;
+    elf_file_header.e_type = ET_CORE;
+    elf_file_header.e_machine = EM_386;
+    elf_file_header.e_version = 1;
+    elf_file_header.e_entry = 0;
+    elf_file_header.e_phoff = sizeof(Elf32_Ehdr);
+    elf_file_header.e_shoff = 0;
+    elf_file_header.e_flags = 0;
+    elf_file_header.e_ehsize = sizeof(Elf32_Ehdr);
+    elf_file_header.e_shentsize = sizeof(Elf32_Shdr);
+    elf_file_header.e_phentsize = sizeof(Elf32_Phdr);
+    elf_file_header.e_phnum = m_regions.size() + 1; // +1 for NOTE section
+    elf_file_header.e_shnum = 0;
+    elf_file_header.e_shstrndx = SHN_UNDEF;
+
+    fd->write(reinterpret_cast<uint8_t*>(&elf_file_header), sizeof(Elf32_Ehdr));
+
+    // TODO: We need to generate appropriate PT_NOTE sections to store
+    // register information and process status stuff here. This is done
+    // by creating a PT_NOTE Phdr for every threa and filling
+    // in the appropriate fields.
+
+    // Dump all regions to file
+    for (Region& region : m_regions) {
+        Elf32_Phdr phdr;
+        VMObject& vmobj = region.vmobject();
+
+        phdr.p_type = PT_LOAD;
+        phdr.p_offset = offset;
+        phdr.p_vaddr = reinterpret_cast<uint32_t>(region.vaddr().as_ptr());
+        phdr.p_paddr = 0;
+
+        // This is a bit janky because we need to work out how big the memory
+        // region will be in the file. This sadly can't just be region.size()
+        // because it will mess up the size of the file. GDB can detect, and
+        // WILL complain about this.
+        size_t mapped_page_count = 0;
+        for (size_t page_num = 0; page_num < vmobj.page_count(); page_num++) {
+            PhysicalPage* page = vmobj.physical_pages()[page_num];
+            if (!page)
+                continue;
+
+            mapped_page_count++; // Only consider mapped pages (so we don't go past the size of `region`)
+        }
```

Comment:
This code is a bit wrong. A `Region` is a wrapper around some or all of the pages in a `VMObject`.
If you `mmap()` some memory and then `unmap()` half of it, the `VMObject` may still be the original size, while you only have a half-sized `Region` now.

The correct number of pages to dump is the number of pages in the `Region`. So `Region::page_count()`.

Whether a page is backed by a physical page or not only matters when it comes to grabbing its data. (For unmapped pages, we can simply memset the output with zeroes.)

Reply:
Ah that makes sense, that'd explain what GDB was complaining about. So run the loop over `Region::page_count()` and then if the page isn't mapped (`!page`) just output a bunch of zeroes instead?

Reply:
> if the page isn't mapped (`!page`) just output a bunch of zeroes instead?

`!page` is not about whether the (virtual) page is mapped, it's about whether it's backed by a physical page. It may not have a backing physical page in case it was purged, or never faulted in. In the former case it's fine to dump it as a bunch of zeros, but in the latter case we'd have to fault in the actual page and dump that.

Reply:
These can surely just be ignored then? The whole point of the core dump is really to only dump whatever is actually backed by a physical page at the time of the crash if what I've read is correct. `FileSiz` is the size of the data that is immediately available when something goes wrong.

Reply:
I'm not familiar with how it's stored in the ELF, but I would expect to see all the pages that were logically mapped (mmaped) for the process (even if they were paged out / never paged in).